src/aproxy.py

```diff
@@ -0,0 +1,463 @@
+                raise ValueError(f"invalid port range: {item}") from exc
+            if not 1 <= start <= end <= 65535:
+                raise ValueError(f"port range must be between 1 and 65535 instead of {item}")
+            ranges.append((start, end))
+
+        return ranges
+
+    @classmethod
+    def _get_principal_proxy_address(cls) -> str:
+        """Get proxy address from Juju-provided environment variables.
+
+        Juju automatically exposes proxy configuration to charms as environment
+        variables (JUJU_CHARM_HTTP_PROXY, JUJU_CHARM_HTTPS_PROXY).
+
+        Returns:
+            The proxy address (hostname or IP) if set, else an empty string.
+        """
+        https_proxy = os.environ.get("JUJU_HTTPS_PROXY", "")
+        http_proxy = os.environ.get("JUJU_HTTP_PROXY", "")
+
+        proxy_conf = https_proxy or http_proxy or ""
+
+        # Strip any leading http:// or https://
+        return re.sub(r"^https?://", "", proxy_conf)
+
+
+class AproxyManager:
+    """Manages aproxy snap and nft configuration persistence."""
+
+    def __init__(self, config: AproxyConfig, charm: ops.CharmBase):
+        """Construct.
+
+        Args:
+            config: AproxyConfig instance with current configuration.
+            charm: The charm instance, used to access model bindings.
+        """
+        self.config = config
+        self.charm = charm
+
+    # ---------------- Snap ----------------
+
+    def install(self) -> None:
+        """Install aproxy snap using snap helper."""
+        logger.info("Installing %s snap", APROXY_SNAP_NAME)
+        snap_cache = snap.SnapCache()
+        snap_cache[APROXY_SNAP_NAME].ensure(
+            state=snap.SnapState.Latest, channel=APROXY_SNAP_CHANNEL
+        )
+
+    def uninstall(self) -> None:
+        """Remove aproxy snap using snap helper."""
+        logger.info("Removing %s snap", APROXY_SNAP_NAME)
+        snap_cache = snap.SnapCache()
+        snap_cache[APROXY_SNAP_NAME].ensure(state=snap.SnapState.Absent)
+
+    def is_snap_installed(self) -> bool:
+        """Check if aproxy snap is installed.
+
+        Returns:
+            True if installed, False otherwise.
+        """
+        snap_cache = snap.SnapCache()
+        return snap_cache[APROXY_SNAP_NAME].present
+
+    def configure_target_proxy(self) -> None:
+        """Configure aproxy snap with current config.
+
+        Raises:
+            ConnectionError: If the target proxy is not reachable.
+        """
+        target_proxy = f"{self.config.proxy_address}:{self.config.proxy_port}"
+        if not self._is_proxy_reachable(self.config.proxy_address, self.config.proxy_port):
+            logger.error("Proxy is not reachable at %s", target_proxy)
+            raise ConnectionError(f"Proxy is not reachable at {target_proxy}")
+
+        logger.info("Configuring snap: proxy=%s", target_proxy)
+        snap_cache = snap.SnapCache()
+        snap_cache[APROXY_SNAP_NAME].set({"proxy": target_proxy})
```

Comment:
I think here it should check the current configuration of aproxy, compare it to the desired configuration, and only set the configuration if they differ. As even if the configuration values are the same, setting the configuration will trigger the config hook and cause the aproxy service to restart, which can be avoided.